### PR TITLE
Add glint support

### PR DESCRIPTION
**Pull request by gossi**
This PR:

- [x] converts the addon itself into TS
- [x] Provides [Helper Signature types](https://typed-ember.gitbook.io/glint/environments/ember/helper-and-modifier-signatures)
- [x] Tests keep working

... turns out, the CI wasn't running for a while and this at least has some problems locally - it might be, the current `main` branch is broken, too.

### CODE REVIEW

**Comment by gossi** [comment]
Oh, I should've looked around at first. I found there is also #107 which adds glint support. I took note and add the registry (as I forgot in the first place). Now, we have registry + export from index :)

File: test-app/ember-cli-build.js [hunk 1]
```diff
@@ -15,11 +19,11 @@ module.exports = function (defaults) {
     behave. You most likely want to be modifying `./index.js` or app's build file
   */
 
-  if (app.env === 'test') {
-    app.import('node_modules/ember-source/dist/ember-template-compiler.js', {
-      type: 'test',
-    });
-  }
+  // if (app.env === 'test') {
```
Comment (NullVoxPopuli) [inline test-app/ember-cli-build.js:22]: why comment this out and leave it?

Comment (gossi) [reply test-app/ember-cli-build.js:22]: I didn't know what it does. So I commented it to see if it has an affect. Do you know what it does?

File: ember-element-helper/src/helpers/element.ts [hunk 2]
```diff
@@ -7,14 +7,24 @@ import { ensureSafeComponent } from '@embroider/util';
...
-    this.tagName = UNINITIALIZED;
-    this.componentClass = null;
-  }
+type Positional<T> = [name: T];
+type Return<T> = EmberComponent<{
+  Element: T extends keyof HTMLElementTagNameMap ? HTMLElementTagNameMap[T] : Element;
+  Blocks: { default: [] };
+}>;
+
+export interface ElementSignature<T extends string> {
```
Comment (NullVoxPopuli) [inline ember-element-helper/src/helpers/element.ts:16]: any reason not to use this strategy? https://github.com/tildeio/ember-element-helper/pull/102#discussion_r1186850830

Comment (gossi) [reply ember-element-helper/src/helpers/element.ts:16]: it is using those types. I stripped them away, so I can use individual parts (`Returns` and `Positional`) within the class.

**Comment by NullVoxPopuli** [comment]
Can you make this pr only about the glint types?

Upgrading embroider requires that we have a correct dep graph, and yarn doesn't provide that (nor is it capable)

File: ember-element-helper/package.json [hunk 3]
```diff
@@ -44,9 +52,18 @@
     "@babel/plugin-proposal-class-properties": "7.18.6",
     "@babel/plugin-syntax-decorators": "7.18.6",
     "@babel/plugin-proposal-decorators": "7.18.6",
-    "@embroider/addon-dev": "1.8.3",
+    "@babel/preset-typescript": "7.22.5",
+    "@babel/runtime": "^7.22.6",
+    "@embroider/addon-dev": "3.2.0",
```
Comment (NullVoxPopuli) [inline ember-element-helper/package.json:57]: can we up to addon-dev @v4? :sweat_smile: was just released this morning :partying_face:

**Review by NullVoxPopuli — APPROVED** [review]
This is looking amazing!!!!! thank you!!!!

**Review by NullVoxPopuli — APPROVED** [review]
looks good to me, thanks for working on this!

**Review by NullVoxPopuli — CHANGES_REQUESTED** [review]
oh but wait -- we need to run a typescript matrix in Ci, to assure we have TS compatibility.

I usually do TS 4.8, 4.9, 5.0, and 5.1 when running `glint` on the test-app

here is an example: https://github.com/NullVoxPopuli/ember-resources/blob/main/.github/workflows/ci.yml#L50-L76

File: test-app/app/components/element-receiver.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+interface ElementReceiverSignature<T extends string> {
+  Element: ElementFromTagName<T>;
+  Args: {
+    tag: ElementSignature<T>['Return'];
+  };
+  Blocks: {
+    default: [];
+  }
+}
```
Comment (gossi) [inline test-app/app/components/element-receiver.ts:12]: This is my test type, but they appear not to work perfectly. Here is the glint error message:

```
app/components/element-receiver.hbs:2:3 - error TS2345: Argument of type 'NonNullable<ElementFromTagName<T>> extends never ? unknown : ElementFromTagName<T>' is not assignable to parameter of type 'Element'.
  Type 'unknown' is not assignable to type 'Element'.

2   <Tag id="content" ...attributes>{{yield}}</Tag>
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

app/components/element-receiver.hbs:2:21 - error TS2345: Argument of type 'NonNullable<ElementFromTagName<T>> extends never ? unknown : ElementFromTagName<T>' is not assignable to parameter of type 'Element'.
  Type 'unknown' is not assignable to type 'Element'.

2   <Tag id="content" ...attributes>{{yield}}</Tag>
```

File: ember-element-helper/README.md [hunk 5]
```diff
@@ -85,12 +80,93 @@ that accepts "contextual components" as arguments:
...
+
+interface YourComponentSignature<
+  T extends string = 'section'
+> {
+  Element: HTMLButtonElement | HTMLAnchorElement | ElementFromTagName<T>;
+  Args: {
+    element?: ElementSignature<T>['Return'];
+  };
+}
+```
```
Comment (gossi) [inline ember-element-helper/README.md:164]: This is giving me problems. See below: https://github.com/tildeio/ember-element-helper/pull/107/files#r1286107261

File: ember-element-helper/README.md [hunk 6]
```diff
@@ -85,12 +80,93 @@ that accepts "contextual components" as arguments:
...
+declare module '@glint/environment-ember-loose/registry' {
+  export default interface Registry extends EmberElementHelperRegistry, /* other addon registries */ {
+    // local entries
+  }
+}
+```
+
+> **Note:** Glint itself is still under active development, and as such breaking changes might occur.
+> Therefore, Glint support by this addon is also considered experimental, and not covered by our SemVer contract!
+
```
Comment (gossi) [inline ember-element-helper/README.md:119]: Types docs start here. This is giving me problems. See below: https://github.com/tildeio/ember-element-helper/pull/107/files#r1286107261

**Comment by gossi** [comment]
Typechecks for TS 4.8+ with Glint. Had to use `{{!@glint-ignore}}` due to a found, which is documented at typed-ember/glint#610.

**Review by NullVoxPopuli — APPROVED** [review]
Gonna try to get a review from the TS folks as well -- since the impact of this library is so wide

**Comment by gossi** [comment]
The problem is more likely to sit in upstream `@glint/template`. I tested the suggested change (as per draft PR: typed-ember/glint#613) locally (see test protocol in linked PR).

**Pull request merged**